ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+    {
+        FILE_TYPE *f_cpg;
+
+        strcpy(cpg_file, pMMBDXP->szFileName);
+        strcpy(cpg_file, reset_extension(cpg_file, "cpg"));
+        f_cpg = fopen_function(cpg_file, "r");
+        if (f_cpg)
+        {
+            char *p;
+            size_t read_bytes;
+            fseek_function(f_cpg, 0L, SEEK_SET);
+            if (11 > (read_bytes = fread_function(local_message, 1, 10, f_cpg)))
+            {
+                local_message[read_bytes] = '\0';
+                p = strstr(local_message, "UTF-8");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_UTF8_DBF;
+                p = strstr(local_message, "UTF8");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_UTF8_DBF;
+                p = strstr(local_message, "ISO-8859-1");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_ANSI_DBASE;
+            }
+            fclose_function(f_cpg);
+        }
+    }
+    if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+    {
+        memcpy(&pMMBDXP->FirstRecordOffset, &offset_primera_fitxa, 2);
+        memcpy(((char *)&pMMBDXP->FirstRecordOffset) + 2, &pMMBDXP->reserved_2,
+               2);
+
+        if (n_queixes_estructura_incorrecta > 0)
+            offset_fals = pMMBDXP->FirstRecordOffset;
+
+        memcpy(&pMMBDXP->BytesPerRecord, &ushort, 2);
+        memcpy(((char *)&pMMBDXP->BytesPerRecord) + 2, &pMMBDXP->reserved_1, 2);
+    }
+    else
+    {
+        pMMBDXP->FirstRecordOffset = offset_primera_fitxa;
+        pMMBDXP->BytesPerRecord = ushort;
+    }
+
+    /* ====== Record structure ========================= */
+
+    if (n_queixes_estructura_incorrecta > 0)
+        pMMBDXP->nFields = (MM_EXT_DBF_N_FIELDS)(((offset_fals - 1) - 32) / 32);
+    else
+    {
+        MM_ACUMULATED_BYTES_TYPE_DBF bytes_acumulats = 1;
+
+        pMMBDXP->nFields = 0;
+
+        fseek_function(pf, 0, SEEK_END);
+        if (32 < (ftell_function(pf) - 1))
+        {
+            fseek_function(pf, 32, SEEK_SET);
+            do
+            {
+                bytes_per_camp = 0;
+                fseek_function(
+                    pf,
+                    32 + (MM_FILE_OFFSET)pMMBDXP->nFields * 32 +
+                        (MM_MAX_LON_CLASSICAL_FIELD_NAME_DBF + 1 + 4),
+                    SEEK_SET);
+                if (1 != fread_function(&bytes_per_camp, 1, 1, pf) ||
+                    1 != fread_function(&un_byte, 1, 1, pf) ||
+                    1 != fread_function(&tretze_bytes,
+                                        3 + sizeof(bytes_per_camp), 1, pf))
+                {
+                    free(pMMBDXP->pField);
+                    fclose_function(pf);
+                    return 1;
+                }
+                if (bytes_per_camp == 0)
+                    memcpy(&bytes_per_camp, (char *)(&tretze_bytes) + 3,
+                           sizeof(bytes_per_camp));
+
+                bytes_acumulats += bytes_per_camp;
```

Comment:
Given that bytes_acumulats is a GUInt32 and bytes_per_camp too, and that bytes_per_camp is not validated, there's a chance that bytes_acumulats could overflow. As those variables are unsigned, this is not a undefined behaviour, but this could make the logic wrong. It might be simpler to promote bytes_acumulats to a GUInt64

Reply:
done